namespace/joiner.go

```diff
@@ -0,0 +1,406 @@
+	fd        int             // The file descriptor of the namespace.
+	flags     uintptr         // The flags to use when joining the namespace.
+
+	isJoined bool // A boolean to indicate if the namespace has been joined.
+}
+
+// ReverseOrder returns a reversed copy of the JoinerContexts.
+func (joiners JoinerContexts) ReverseOrder() JoinerContexts {
+	reversed := make(JoinerContexts, len(joiners))
+	for i, j := len(joiners)-1, 0; i >= 0; i, j = i-1, j+1 {
+		reversed[j] = joiners[i]
+	}
+	return reversed
+}
+
+// JoinReverse joins all the namespaces in the JoinerContexts in reverse order.
+func (joiners *JoinerContexts) JoinReverse() (err error) {
+	*joiners = joiners.ReverseOrder()
+	return joiners.Join()
+}
+
+// Join joins all the namespaces in the JoinerContexts.
+func (joiners *JoinerContexts) Join() (err error) {
+	for _, joiner := range *joiners {
+		if joiner.isJoined {
+			logrus.Tracef("Already joined namespace: %s", joiner.namespace)
+			continue
+		}
+
+		if joiner.namespace == types.NamespaceMnt {
+			err := unix.Unshare(unix.CLONE_NEWNS)
+			if err != nil {
+				return errors.Wrapf(err, "failed to unshare namespace: %+s", joiner.namespace)
+			}
+		}
+
+		if err := unix.Setns(joiner.fd, 0); err != nil {
+			return errors.Wrapf(err, "failed to set namespace: %+s", joiner.namespace)
+		}
+
+		joiner.isJoined = true
+		logrus.Tracef("Joined namespace: %v", joiner.namespace)
+	}
+	return nil
+}
+
+// Reset resets all the JoinerContexts.
+func (joiners *JoinerContexts) Reset() (err error) {
+	for _, joiner := range *joiners {
+		logrus.Tracef("Resetting namespace: %+v", joiner)
+		joiner = &JoinerContext{}
+	}
+	return nil
+}
+
+// OpenFile opens a file in the JoinerContexts.
+func (joiners *JoinerContexts) OpenFile(path string) (fd int, err error) {
+	return unix.Open(path, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+}
+
+// CloseFiles closes all the files in the JoinerContexts.
+func (joiners *JoinerContexts) CloseFiles() {
+	for _, joiner := range *joiners {
+		if joiner.fd == -1 {
+			continue
+		}
+
+		if err := unix.Close(joiner.fd); err != nil {
+			logrus.WithError(err).Warnf("Failed to close %v namespace file", joiner.namespace)
+			continue
+		}
+
+		joiner.fd = -1
+	}
+}
+
+// Joiner is a struct that holds information about the namespaces to join.
+type Joiner struct {
+	namespaces []types.Namespace // List of namespaces to be joined.
+
+	directory string // The target directory where the namespace files are located.
+	pid       uint64 // Process ID associated with the Joiner.
+
+	origin JoinerContexts // Contexts of the original namespaces.
+	target JoinerContexts // Contexts of the namespaces to be joined.
+
+	stop    chan struct{} // Channel to signal stopping the execution of the function within namespaces.
+	timeout time.Duration // Timeout duration for the execution of the function within namespaces.
+
+	mutex sync.Mutex // Mutex to protect the current access to the Joiner.
```

Comment:
Initially, I was thinking about reusing the Joiner struct to run the functions. After considering the complexity of the implementation, it's now a single dedicated object for the RunFunc to run functions in the namespaces. 

Therefore, I don't think the lock is necessary. This is because all access to the receiver object happens within a single goroutine, which is locked to an OS thread.